### ISSUE TITLE
GeoIP with Round Robin

### ISSUE THREAD

**Issue by ghost** [gdnsd/gdnsd#171]
Hi, love the software and thank you for putting time into it.  Here's what I am trying to do, maybe you can point me in the right direction.

I would like gdnsd to follow suit to geo-ip primarily and then defer to Round Robin after say there is a higher amount of traffic coming in.  Maybe weight the geo-ip plugin with a RR method.

Basically I'm trying to disperse the load to all the cluster nodes in a RR sense, but still put a priority on Geo location if that is possible.

Thank you.

**Comment by xlightwaverx** [comment]
Example:  When traffic is hitting a node, and nginx gets terminated, I have it setup to auto fall back to the other node in the cluster.  (fails on /nginx_status check) What the application does, even though journalctl is showing on both machines that the state of that node is DOWN, is that it keeps sending traffic to the down node, despite confirming in the journal log that it has marked that node as down on both servers.  When doing load testing as well, and doing the complete opposite, as in having the load test with nginx marked as down, starting the test, and then reenabling nginx, the journalctl shows the node is up on both servers, but the traffic doesn't get routed to the new node because it technically is closer to the load testing server. 

Everything appears to be working correctly log wise, but I'm not seeing it in real world testing.  Any help would be greatly appreciated.  

![image](https://user-images.githubusercontent.com/45115802/49161813-9a0a2480-f2f7-11e8-829d-6ecba5545f66.png)

**Comment by blblack** [comment]
It's hard for me to parse this down into terms I can understand concretely!

On "Round Robin" in the DNS in general:

1) "Round robin" can mean (at least) a couple of different things at the DNS level: serving a set of multiple addresses together in different rotated/randomized orders each time, or serving a single IP from a set and rotating/randomizing which singular IP you deliver per-response.  I suspect in this context you're talking about the latter sort of technique.

2) "Round robin" sorts of techniques via DNS aren't generally very helpful for load balancing in practice.  There are too many ways that caches, stub resolvers, and end-clients variously defeat such aims.  With the reordering-multi-address-outputs variant, all kinds of things go wrong in intermediary caches (esp wrt to how many clients end up following a given cache entry from the global space of all client caches), stub resolvers, and end-clients that effectively defeat/bias/distort the intended load distribution (including stub resolvers doing deterministic sorting of the set back into an unshuffled state!).  The rotation of a singular response address from a configured set works a little better against the latter sort of problems with unshuffling, but not against the cache entry biasing problems.

On your specific question about DOWN nodes being ignored in gdnsd routing: this may be a matter of threshold configurations?  It's hard to say without the complete plugin configuration, but in general gdnsd has the notion of thresholding, where if more than a threshold amount of results have failed into the DOWN state, they're all considered up (the idea being that past some point of multiple failure, it's better to assume it's a global problem or monitoring itself has failed, and thus revert to making decisions as if they're all UP).

Integrating feedback from your destination nodes' realtime load is tricky, I don't think there's a good way to do that at present, if that's what you mean (having some kind of modal switch to even distribution as load increases).

Are you using geoip to balance into a local cluster, or for true geographic endpoints?  Are you just wanting some protection against overloading one location when another is relatively-idle?

**Comment by xlightwaverx** [comment]
Thanks for the prompt reply.  Here is the testing configuration.  I have lowered the thresholds to what it allows me to without failing checkconf.

```
options => {
  username => nobody
  }

service_types => {
  servicecluster => { plugin => http_status, vhost => localhost, url_path => "/nginx_status", down_thresh => 1, interval => 2, ok_thresh => 1,up_thresh => 1 }
}
plugins => {
  geoip => {
    maps => {
      auto_map => {
        geoip2_db = GeoLite2-City.mmdb,
        datacenters => [dc1, dc2],
        auto_dc_coords => {
          dc1 => [ 60.1708, 24.9375 ],
                    dc2 => [ 51.2993, 9.491 ],
                  },
        # only fail through the nearest 2 before giving up:
        auto_dc_limit => 2,
      }
      $include{geoip_maps}
    },
    resources => {
      ip0 => {
        map => auto_map,
        service_types => [ servicecluster ],
        dcmap => {
          dc1 => x.x.x.x,
                    dc2 => x.x.x.x,
                  }
      },
      $include{geoip_resources}
    }
  },
  metafo => {
    resources => {
      ip0 => {
        datacenters => [ real ],
        dcmap => {
          real => %geoip!ip0
        }
      },
      $include{metafo_resources}
    }
  }
}
```

> Are you using geoip to balance into a local cluster, or for true geographic endpoints? Are you just wanting some protection against overloading one location when another is relatively-idle?

Yes this is the issue I am trying to fix.  Using bind alone with A entries to both IP addresses on both nodes of the cluster leads to a balanced enough approach that I am happy with.  The RR loadtesting fills the pipes almost equally (Not 100% level but close enough).

I am trying to incorporate gdnsd into the mix so I can:

1) Add GeoIP location so I can send the client to the closest node.
2) Yes, I do not want that one load overloaded.
3) I would like to add a weighted effect to the config if I could, something that would try GeoIP first, as if a customer was accessing the site, they simply should be provided the closest node.
4) If things get busy, like maxing out the pipe busy, I would like to revert back to a RR setup where the load is dispersed again.

The end goal is to have HA with speed.  With the current setup, the load is dispersed evenly.  If a node goes down, clients are no longer being sent to the down node. (If they are, the browser tries the next A entry in line, which, besides a brief pause, leads the browser to the node with the active site on, as if nothing was really down.

But if I can add Gdnsd to the mix, it allows new features.  

It can detect if the webserver is down.  In the current setup of RR I use, if the webserver is down, visitors are still sent to the IP of the "somewhat down" node, leading to a failure.  Gdnsd can solve that, I already figured that out in my short testing of it.  If a node goes down with Gdnsd, gdnsd is smart enough to not send visitors to the down node.  But in my loadtesting using gdnsd, I have just found that even though journalctl shows when the node goes down, and back up during the testing process, it still doesn't redirct the user.  This may be the load testing tool as well simply sticking to a cached IP address.
  
So basically, if possible, what I am trying to do is have at least a first check for GeoIP, send the visitor to the correct node, and after RR would be fine to distribute load.  

If it's not possible that's fine, that's what testing is for :)

**Comment by blblack** [comment]
> If a node goes down with Gdnsd, gdnsd is smart enough to not send visitors to the down node. But in my loadtesting using gdnsd, I have just found that even though journalctl shows when the node goes down, and back up during the testing process, it still doesn't redirct the user. This may be the load testing tool as well simply sticking to a cached IP address.

I'm going to assume this is some caching in the loadtesting tool for now.  I don't see why this wouldn't work as advertised!
 
> So basically, if possible, what I am trying to do is have at least a first check for GeoIP, send the visitor to the correct node, and after RR would be fine to distribute load.
> [and earlier] If things get busy, like maxing out the pipe busy, I would like to revert back to a RR setup where the load is dispersed again.

I still don't follow the desired logic here.  Can you explain it in some kind pseudo-code involving healthcheck states and/or load states (or is overload determined by failing the same healthcheck?)

**Comment by xlightwaverx** [comment]
> I still don't follow the desired logic here. Can you explain it in some kind pseudo-code involving healthcheck states and/or load states (or is overload determined by failing the same healthcheck?)

I guess it would depend on the threshold setup.  If a node fails to respond, say after a second, or two, enable the multifo or metafo mode possibly?  I believe multifo or metafo are what handles the round robin mode correct, or not round robin rather, but just gives all nodes as available and the browser does what it wants, correct?

And if that's not doable,

A load detection setup would suffice.  But yes, overload would add lots of latency and if that can be measured via healthcheck, that would suffice I believe.  The more taxing php-fpm and nginx are on the node, the more latency tcp/ip traffic would have, and if the node fails, drop the node.  Ideally, switching to a round robin setup would be the best method if that's possible with this software, but deflecting that node until the healthcheck passes would do the same thing.

**Comment by blblack** [comment]
> Ideally, it would start with geolocation as a priority if this is weighable. I guess it would depend on the threshold setup. If a node fails to respond, say after a second, or two, enable the multifo or metafo mode possibly? I believe multifo or metafo are what handles the round robin mode correct, or not round-robin rather, but just gives all nodes as available and the browser does what it wants, correct?

Yeah I still don't get the logic, or what I do get of it doesn't seem to make much sense to me.  It seems like you're saying this, in pseudo-code terms:
```
if (both_nodes_healthy()) {
    route_via_geoip();
} else {
    // one or both nodes are sick by healthcheck
    serve_both_addrs_roundrobin();
}
```

Note the default behavior of the geoip plugin, with just 2 DCs defined with 1 IP each, and no strange tuning of the healthcheck or or other tunables, would be:

```
if (both_nodes_healthy()) {
    route_via_geoip();
} else if (one_node_sick()) {
    all_traffic_to_healthy_node();
} else {
    // both nodes sick
    route_via_geoip();
}
```

The key distinction from the default behavior to my guess at the behavior you want, is that in the situation that 1/2 nodes is detected as sick, you seem to want to keep the sick address in use?  The final option for what to do when both nodes are sick is easier to deal with (metafo could chose all-address output over geoip routing in that case, but I don't see why that would be advantageous over the default geoip routing here).

In general, it's much easier to deal with geographic balancing/distribution if you can start with the assumption that all your allowed failure modes are load-tolerable.  With only 2 sites in play, you have two policy options here that fit: either decide that each site is built to withstand expected total global load, or decide that neither site is allowed to go offline (which probably isn't reasonable, as even with the best site-local redundancy you're probably going to have issues with reachability and/or maintenance from time to time).

**Comment by blblack** [comment]
No followup, not sure what to do here (although it's possible I'm missing something that leads to a feature request here!)
